sagemaker-neo-notebooks/edge/cpp-integration/tutorial.cc

```diff
@@ -0,0 +1,622 @@
+#include <chrono>
+#include <ctime>
+#include <exception>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+// 3rd-party
+#include <nlohmann/json.hpp>
+#include "npy.hpp"
+
+// aws-sdk
+#include <aws/core/Aws.h>
+#include <aws/core/client/AWSError.h>
+#include <aws/iam/IAMClient.h>
+#include <aws/iam/model/AttachRolePolicyRequest.h>
+#include <aws/iam/model/CreateRoleRequest.h>
+#include <aws/iam/model/GetRoleRequest.h>
+#include <aws/iam/model/ListRolesRequest.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/CreateBucketRequest.h>
+#include <aws/s3/model/GetObjectRequest.h>
+#include <aws/s3/model/ListObjectsRequest.h>
+#include <aws/s3/model/PutObjectRequest.h>
+#include <aws/sagemaker/SageMakerClient.h>
+#include <aws/sagemaker/model/CreateCompilationJobRequest.h>
+#include <aws/sagemaker/model/DescribeCompilationJobRequest.h>
+#include <aws/sagemaker/model/Framework.h>
+#include <aws/sagemaker/model/InputConfig.h>
+
+// internal
+#include <dlr.h>
+
+using namespace std;
+using json = nlohmann::json;
+
+const string ROLE_NAME = "windows-demo-test-role";
+
+Aws::S3::S3Client getS3Client() {
+  // for tutorial, we set region to us-west-2
+  const string region = "us-west-2";
+  Aws::String region_aws_s(region.c_str(), region.size());
+
+  // init client
+  Aws::Client::ClientConfiguration client_config;
+  client_config.region = region_aws_s;
+  Aws::S3::S3Client s3_client(client_config);
+  return s3_client;
+}
+
+Aws::IAM::IAMClient getIamClient() {
+  // create IAM role, note that IAM only support global region.
+  Aws::IAM::IAMClient iam_client;
+  return iam_client;
+}
+
+Aws::SageMaker::SageMakerClient getSageMakerClient() {
+  // for tutorial, we set region to us-west-2
+  const string region = "us-west-2";
+  Aws::String region_aws_s(region.c_str(), region.size());
+
+  // init client
+  Aws::Client::ClientConfiguration client_config;
+  client_config.region = region_aws_s;
+  Aws::SageMaker::SageMakerClient sm_client(client_config);
+  return sm_client;
+}
+
+template <typename T>
+std::string getErrorMessage(Aws::Client::AWSError<T> error) {
```

Comment:
`error` should be passed by reference. Same for other APIs. Let's pass by reference when we can so that no copy is needed

Reply:
it's revised as suggested.